### PR TITLE
feat: gas oracle

### PR DESCRIPTION
**Pull request by kaladinlight**
## Description

- use unchained gas oracle for slow/average/fast estimation on both eip1559 and legacy fees
- lift approval/fee/build/broadcast functionality to common `utils/evm` which can be shared by defi opportunities and investor packages next (same tx building logic throughout)
- refactor existing swappers to utilize common evm utils, handle less than min quotes better, and general cleanup

Changes:
- all routes should show `insufficient sell amount` with estimated gas and receive value of 0 now (button shows min amount required)
- cowswap will not default to min amount when sell amount is too low to better display the right ratio diffs

NOTE:
unchained will need to be deployed with new gas oracle. this will be a blocking deploy on the backend to ensure enough time for testing and validating the gas oracle works as expected and continues to return accurate fees. once this is confirmed, we can deploy unchained to prod and release this feature as well (testing will be done in yeet)

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Risk

Med/High

## Testing

- Full regression test of all evm swappers
  - ensure 0 value quotes display estimated gas where applicable
  - ensure less than min quotes display gas and exchange rate, but show `Insufficient sell amount` and `Minimum` as expected
  - ensure expected swappers still show for the various trade pairs
  - ensure gas estimations on trade confirm are accurate and very close to confirmed tx fee
  - ensure we can still trade and don't hit any errors through full trade flow

### Engineering

:point_up: 

### Operations

:point_up: 

### Thorchain
Current             |  New
:-------------------------:|:-------------------------:
image | image
image | image
image | image
image | image

image

### CowSwap
Current             |  New
:-------------------------:|:-------------------------:
image | image
image | image
image | image
image | image

image

### Zrx
Current             |  New
:-------------------------:|:-------------------------:
image | image
image | image
image | image
image | image

image

### OneInch
Current             |  New
:-------------------------:|:-------------------------:
image | image
image | image
image | image
image | image

image

### CODE REVIEW

File: src/components/Trade/Components/TradeQuotes/TradeQuote.tsx
Comment (gomesalexandre) [inline src/components/Trade/Components/TradeQuotes/TradeQuote.tsx:212]: This doesn't hold true anymore now with the Oracle, correct?

File: src/components/Trade/TradeInput.tsx
Comment (gomesalexandre) [inline src/components/Trade/TradeInput.tsx:560]: Ditto, does the Gas oracle change means we should never show the "Or Invalid Trade Pair" state for 1inch? At first glance this seems unrelated

File: src/lib/swapper/swappers/CowSwapper/CowSwapper.ts
Comment (gomesalexandre) [inline src/lib/swapper/swappers/CowSwapper/CowSwapper.ts:92]: Might as well 🧠

File: src/lib/swapper/swappers/CowSwapper/types.ts
Comment (gomesalexandre) [inline src/lib/swapper/swappers/CowSwapper/types.ts:30]: Prefer upper snake case for consts but meh

File: src/lib/swapper/swappers/CowSwapper/utils/constants.ts
Comment (gomesalexandre) [inline src/lib/swapper/swappers/CowSwapper/utils/constants.ts:7]: `BY_CHAIN_ID` ?

File: src/lib/swapper/swappers/ZrxSwapper/getZrxTradeQuote/getZrxTradeQuote.ts
Comment (gomesalexandre) [inline src/lib/swapper/swappers/ZrxSwapper/getZrxTradeQuote/getZrxTradeQuote.ts:109]: Is this the same behavior as before? We were getting the `buyAmount` directly from `data` previously

File: src/lib/swapper/swappers/ZrxSwapper/types.ts
Comment (gomesalexandre) [inline src/lib/swapper/swappers/ZrxSwapper/types.ts:5]: Ditto casing

File: src/lib/utils/evm/abis/erc20.ts
Comment (gomesalexandre) [inline src/lib/utils/evm/abis/erc20.ts:1]: We already have the ERC20 ABI in `contracts/abis` (as JSON, though there's generated typegen ABI we might be able to use if types are compatible)

File: src/lib/swapper/swappers/OneInchSwapper/getTradeQuote/getTradeQuote.ts
Comment (gomesalexandre) [inline src/lib/swapper/swappers/OneInchSwapper/getTradeQuote/getTradeQuote.ts:32]: Are we now missing that check?

**Review by gomesalexandre — COMMENTED** [review]
A few early comments, reviewing this as code-only as I'm unable to test this - spotted a few cases of checks that may now be missing / logic that seems to have been changed while unrelated to the gas oracle changes.
Will make a proper second pass with more context and while the gas oracle is testable

File: src/components/Trade/Components/TradeQuotes/TradeQuote.tsx
Comment (kaladinlight) [reply src/components/Trade/Components/TradeQuotes/TradeQuote.tsx:212]: We would have been able to get gas fees regardless, just weren't. So not directly related, but it is using data from the gas oracle now as well

File: src/components/Trade/TradeInput.tsx
Comment (kaladinlight) [reply src/components/Trade/TradeInput.tsx:560]: Reason for these two changes is that the one inch quote wasn't handling sell amounts below min gracefully. It now is so we can display in the same manner as the other swappers.

File: src/lib/swapper/swappers/CowSwapper/types.ts
Comment (kaladinlight) [reply src/lib/swapper/swappers/CowSwapper/types.ts:30]: all of the other ones are the same casing so more following pattern than anything. would rather update all if we want at the same time, but not for this PR with everything else

File: src/lib/swapper/swappers/ZrxSwapper/getZrxTradeQuote/getZrxTradeQuote.ts
Comment (kaladinlight) [reply src/lib/swapper/swappers/ZrxSwapper/getZrxTradeQuote/getZrxTradeQuote.ts:109]: Not the same, we will use `data.buyAmount` if over the minimum amount, otherwise set it to `0` to imply you are unable to trade because it is less than min sell amount.

Before -> After example:
image

Comment (kaladinlight) [reply src/lib/swapper/swappers/ZrxSwapper/getZrxTradeQuote/getZrxTradeQuote.ts:109]: There would need to be additional data passed to the components to show `Insufficient sell amount` and also retain the buy amount returned from the quote, so standardizing current behavior with `0` buy amount when less than min for now

File: src/lib/swapper/swappers/OneInchSwapper/getTradeQuote/getTradeQuote.ts
Comment (kaladinlight) [reply src/lib/swapper/swappers/OneInchSwapper/getTradeQuote/getTradeQuote.ts:32]: Replaced with 
image

File: src/lib/utils/evm/abis/erc20.ts
Comment (kaladinlight) [reply src/lib/utils/evm/abis/erc20.ts:1]: 407142e6ac626e742d64642f320d050971f6f70c

File: src/lib/swapper/swappers/CowSwapper/getCowSwapTradeQuote/getCowSwapTradeQuote.ts
Comment (gomesalexandre) [inline src/lib/swapper/swappers/CowSwapper/getCowSwapTradeQuote/getCowSwapTradeQuote.ts:35]: 👀 
```suggestion
): Promise<Result<TradeQuote<CowChainId, boolean>, SwapErrorRight>> {
```

File: src/lib/swapper/swappers/LifiSwapper/utils/getNetworkFeeCryptoBaseUnit/getNetworkFeeCryptoBaseUnit.ts
Comment (gomesalexandre) [inline src/lib/swapper/swappers/LifiSwapper/utils/getNetworkFeeCryptoBaseUnit/getNetworkFeeCryptoBaseUnit.ts:65]: Now that lib is in web, and that this is now an ethers contract, we should be able to add the abi in `contractManager` and get typechain types for this so we can remove the cast, but let's not tackle in this PR

File: src/lib/swapper/swappers/OneInchSwapper/getTradeQuote/getTradeQuote.ts
Comment (gomesalexandre) [inline src/lib/swapper/swappers/OneInchSwapper/getTradeQuote/getTradeQuote.ts:56]: Do we need the cast? tsc seems happy without

File: src/lib/swapper/swappers/OneInchSwapper/utils/helpers.ts
Comment (gomesalexandre) [inline src/lib/swapper/swappers/OneInchSwapper/utils/helpers.ts:57]:
```suggestion
        message: `[OneInchSwapper: assertValidTrade] - unsupported chainId`,
```

**Review by gomesalexandre — COMMENTED** [review]
A few additional smolish comments but LGTM, tested in a poile. Retested against prod (develop isn't reliable as unchained gas oracle fix is currently deployed for unchained) as a sanity check:

### Minimums are still handled

#### Prod

image
image
image
image


#### This diff

image
image
image
image

### Gas fees and quote slippage/downside are showing no regressions

#### Prod

image
image
image

#### This diff

image
image
image
image
image

Note on the second screenshot in this diff how we now let the trade go through, and the more accurate gas fee

However, when trying to actually broadcast the Tx, I was met with this errror:

image

### 0 amounts are properly handled

#### Prod

image

#### This diff

image

Note missing 1inch because of current 403s, as we've discussed this isn't a bug, we legitimely can't get a 1inch Quote here

Haven't tested trades actually going through, to be tested by ops, but apart from that ZRX Optimism bug, LGTM overall

File: src/lib/swapper/swappers/LifiSwapper/utils/getNetworkFeeCryptoBaseUnit/getNetworkFeeCryptoBaseUnit.ts
Comment (kaladinlight) [reply src/lib/swapper/swappers/LifiSwapper/utils/getNetworkFeeCryptoBaseUnit/getNetworkFeeCryptoBaseUnit.ts:65]: Good point, also contemplated exposing this in unchained since we are already doing it there as well. I wish the node just had a whitelist address to call estimate gas that acts like it has unlimited funds so we can always do full node estimations even on quote haha

File: src/lib/swapper/swappers/OneInchSwapper/getTradeQuote/getTradeQuote.ts
Comment (kaladinlight) [reply src/lib/swapper/swappers/OneInchSwapper/getTradeQuote/getTradeQuote.ts:56]: you're right, was thinking it would give us type safety, but that's not how as works haha. might pull it back out into it's own variable

File: src/lib/swapper/swappers/CowSwapper/getCowSwapTradeQuote/getCowSwapTradeQuote.ts
Comment (kaladinlight) [reply src/lib/swapper/swappers/CowSwapper/getCowSwapTradeQuote/getCowSwapTradeQuote.ts:35]: :brain:

File: src/lib/swapper/swappers/ThorchainSwapper/getThorTradeQuote/getTradeQuote.ts
Comment (0xApotheosis) [inline src/lib/swapper/swappers/ThorchainSwapper/getThorTradeQuote/getTradeQuote.ts:93]:
```suggestion
        message: '[getThorTradeQuote]: receiveAddress is required',
```

File: src/lib/swapper/swappers/ZrxSwapper/utils/helpers/helpers.ts
Comment (0xApotheosis) [inline src/lib/swapper/swappers/ZrxSwapper/utils/helpers/helpers.ts:101]:
```suggestion
        message: '[zrx: getAdapter] - invalid chain adapter',
```

**Review by 0xApotheosis — COMMENTED** [review]
Noticed a few trivial things whilst going through, nothing I'd consider a blocker.
All monads appear to be unwrapped before accessing 👌 

I'll give a more functional test, too.

**Review by 0xApotheosis — APPROVED** [review]
Tested quotes against prod, all is looking well.

Some slight (more correct) changes compared to prod with respect to ordering of swappers for some quotes, but that is expected as we've changed the logic to we don't need to fetch dummy quotes to get gas.

Did not execute trades, will leave that to ops to confirm before merging to `develop`.

File: src/lib/swapper/swappers/ThorchainSwapper/getThorTradeQuote/getTradeQuote.ts
Comment (0xApotheosis) [reply src/lib/swapper/swappers/ThorchainSwapper/getThorTradeQuote/getTradeQuote.ts]: https://github.com/shapeshift/web/pull/4769/commits/e8355d490c6f8e80933ff34b95fa5cc460f034a2

File: src/lib/swapper/swappers/ZrxSwapper/utils/helpers/helpers.ts
Comment (0xApotheosis) [reply src/lib/swapper/swappers/ZrxSwapper/utils/helpers/helpers.ts]: https://github.com/shapeshift/web/pull/4769/commits/e8355d490c6f8e80933ff34b95fa5cc460f034a2

File: src/lib/swapper/swappers/OneInchSwapper/getTradeQuote/getTradeQuote.test.ts
Comment (0xApotheosis) [reply src/lib/swapper/swappers/OneInchSwapper/getTradeQuote/getTradeQuote.test.ts]: https://github.com/shapeshift/web/pull/4769/commits/e8355d490c6f8e80933ff34b95fa5cc460f034a2

File: src/lib/swapper/swappers/ZrxSwapper/zrxBuildTrade/zrxBuildTrade.ts
Comment (0xApotheosis) [reply src/lib/swapper/swappers/ZrxSwapper/zrxBuildTrade/zrxBuildTrade.ts]: https://github.com/shapeshift/web/pull/4769/commits/e8355d490c6f8e80933ff34b95fa5cc460f034a2

File: src/lib/swapper/swappers/CowSwapper/getCowSwapTradeQuote/getCowSwapTradeQuote.ts
Comment (0xApotheosis) [reply src/lib/swapper/swappers/CowSwapper/getCowSwapTradeQuote/getCowSwapTradeQuote.ts]: https://github.com/shapeshift/web/pull/4769/commits/e8355d490c6f8e80933ff34b95fa5cc460f034a2

**Review by gomesalexandre — APPROVED** [review]
Stamp after latest cleanups, note this has been ops tested and ready to go in develop/prod as soon as unchained is deployed

**Pull request merged**
